plugins/SlicerT/SlicerT.cpp

```diff
@@ -0,0 +1,445 @@
+	, m_enableSync(true, this, tr("BPM sync"))
+	, m_originalSample()
+	, m_phaseVocoder()
+	, m_parentTrack(instrumentTrack)
+{
+	m_sliceSnap.addItem("Off");
+	m_sliceSnap.addItem("1/1");
+	m_sliceSnap.addItem("1/2");
+	m_sliceSnap.addItem("1/4");
+	m_sliceSnap.addItem("1/8");
+	m_sliceSnap.addItem("1/16");
+	m_sliceSnap.addItem("1/32");
+	m_sliceSnap.setValue(0); // no snap by default
+
+	m_resamplerState = src_new(SRC_SINC_MEDIUM_QUALITY, 2, nullptr); // no error
+}
+
+SlicerT::~SlicerT()
+{
+	src_delete(m_resamplerState);
+}
+
+void SlicerT::playNote(NotePlayHandle* handle, sampleFrame* workingBuffer)
+{
+	if (m_originalSample.frames() < 2048) { return; }
+
+	// playback parameters
+	const int noteIndex = handle->key() - m_parentTrack->baseNote();
+	const int playedFrames = handle->totalFramesPlayed();
+	const fpp_t frames = handle->framesLeftForCurrentPeriod();
+	const f_cnt_t offset = handle->noteOffset();
+	const int bpm = Engine::getSong()->getTempo();
+	const float pitchRatio = pow(2, m_parentTrack->pitchModel()->value() / 1200);
+	const float inversePitchRatio = 1.0f / pitchRatio;
+
+	// update scaling parameters
+	float speedRatio = static_cast<float>(m_originalBPM.value()) / bpm;
+	if (!m_enableSync.value()) { speedRatio = 1; } // disable timeshift
+	m_phaseVocoder.setScaleRatio(speedRatio);
+	speedRatio *= inversePitchRatio; // adjust for pitch bend
+
+	int totalFrames = inversePitchRatio * m_phaseVocoder.frames(); // adjust frames played with regards to pitch
+	int sliceStart, sliceEnd;
+	if (noteIndex > m_slicePoints.size() - 2 || noteIndex < 0) // full sample if ouside range
+	{
+		sliceStart = 0;
+		sliceEnd = totalFrames;
+	}
+	else
+	{
+		sliceStart = m_slicePoints[noteIndex] * speedRatio;
+		sliceEnd = m_slicePoints[noteIndex + 1] * speedRatio;
+	}
+
+	// slice vars
+	int sliceFrames = sliceEnd - sliceStart;
+	int currentNoteFrame = sliceStart + playedFrames;
+	int noteFramesLeft = sliceFrames - playedFrames;
+
+	if (noteFramesLeft > 0)
+	{
+		int framesToCopy = pitchRatio * frames + 1; // just in case
+		int framesIndex = pitchRatio * currentNoteFrame;
+		framesIndex = std::min(framesIndex, m_phaseVocoder.frames() - framesToCopy);
+
+		// load sample segmengt, with regards to pitch settings
+		std::vector<sampleFrame> prePitchBuffer(framesToCopy, {0.0f, 0.0f});
+		m_phaseVocoder.getFrames(prePitchBuffer.data(), framesIndex, framesToCopy);
+
+		// if pitch is changed, resample, else just copy
+		if (!typeInfo<float>::isEqual(pitchRatio, 1.0f))
+		{
+			SRC_DATA resamplerData;
+
+			resamplerData.data_in = (float*)prePitchBuffer.data();	   // wtf
+			resamplerData.data_out = (float*)(workingBuffer + offset); // wtf is this
+			resamplerData.input_frames = prePitchBuffer.size();
+			resamplerData.output_frames = frames;
+			resamplerData.src_ratio = inversePitchRatio;
+
+			src_process(m_resamplerState, &resamplerData);
+		}
+		else { memcpy(workingBuffer + offset, prePitchBuffer.data(), frames * sizeof(sampleFrame)); }
```

Comment:
Odd, I figured @DanielKauss made the ``memcpy`` to ``std::copy_n`` transition already?

Reply:
I did, but I thought that `copy_n` only worked with iterators, and since `workingbuffer` was a c-style array I left it as is. Sorry, my mistake, didn't realize you can also use arrays, I'm not very familiar with c++.

Reply:
Done